packages/synapse-client/src/SynapseClient.ts

```diff
@@ -41,17 +41,63 @@ import { VerificationServicesApi } from './generated/apis/VerificationServicesAp
 import { WebhookServicesApi } from './generated/apis/WebhookServicesApi'
 import { WikiPageServices2Api } from './generated/apis/WikiPageServices2Api'
 import { WikiPageServicesApi } from './generated/apis/WikiPageServicesApi'
-import { BaseAPI, Configuration } from './generated/runtime'
-import { fetchWithExponentialTimeout } from './util/fetchWithExponentialTimeout'
+import { ErrorResponse } from './generated/models/ErrorResponse'
+import {
+  Configuration,
+  ConfigurationParameters,
+  ErrorContext,
+  ResponseContext,
+} from './generated/runtime'
+import { fetchResponseWithExponentialTimeout } from './util/fetchWithExponentialTimeout'
+import { NETWORK_UNAVAILABLE_MESSAGE } from './util/Constants'
+import { SynapseClientError } from './util/SynapseClientError'
 
-const DEFAULT_CONFIG = new Configuration({
-  fetchApi: fetchWithExponentialTimeout,
-})
+const DEFAULT_CONFIG_PARAMETERS: ConfigurationParameters = {
+  fetchApi: fetchResponseWithExponentialTimeout,
+  middleware: [
+    {
+      async post(context: ResponseContext): Promise<Response | void> {
+        const { response, url } = context
+        if (!response.ok) {
+          const error = await response.json()
+          if (
+            error !== null &&
+            typeof error === 'object' &&
+            'reason' in error
+          ) {
+            throw new SynapseClientError(
+              response.status,
+              (error as ErrorResponse).reason!,
+              url,
+              error as ErrorResponse,
+            )
+          } else {
+            throw new SynapseClientError(
+              response.status,
+              JSON.stringify(error),
+              url,
+            )
+          }
+        }
+      },
+      // Convert error objects to our SynapseClientError
+      onError(context: ErrorContext): Promise<Response | void> {
+        const { response, error } = context
+        console.error(error)
+        throw new SynapseClientError(
+          0,
```

Comment:
This looks familiar (like the network failure handling from the old SynapseClient), but I just want to confirm that we are not losing response.status information here

Reply:
I believe this callback only runs when `fetch` throws an error, in which case there is no response object. This should give us identical behavior to the previous implementation.